### PR TITLE
debug:expression command

### LINKED ISSUE #290: Improve developer experience with a CLI command to test selectors (glob)
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no
| BC Break    | no

#### Summary

It would be very nice to have a CLI command to quickly test selectors like that:

```bash
# simple version - only shows matches:
$ phparkitect selector:test namespace *Controller
Matches:
App/Controller/UserController.php
App/Controller/CustomerController.php

# robust version that shows everything 
$ phparkitect selector:test namespace *Controller --show-all
Matches | File
❌      | App/Entity/User.php
✔️      | App/Controller/UserController.php
✔️      | App/Controller/CustomerController.php
❌      | App/Repository/UserRepository.php
```

Inspiration came from Symfony:

 - [Debug @ Symfony Router](https://symfony.com/doc/6.1/routing.html#debugging-routes):
   - `bin/console debug:router name-of-router`
   - `bin/console router:match /a/nice/path`
 - [Debug @ Symfony Service Container](https://symfony.com/doc/6.1/service_container/debug.html)
   - `bin/console debug:container App\Service\Mailer`

### PR DESCRIPTION
**Pull request by fain182**
This should fix #290 

There are some little things to fix, but the core of the functionality should be here...

- [x] fix the error on psalm (or change the psalm configuration)
- [x] add the command in the documentation
- [x] now it uses the current directory as the default classSet, it's a good idea?
- [x] improve error handling when there are too few or too many arguments for the Expression
- [ ] it would be great to add all the expressions in the documentation of the command dinamically
- [x] decide how to manage the php version for the parsing
- [x] decide if we care about parsing errors

### CODE REVIEW

File: src/CLI/Command/DebugExpression.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        $targetPhpVersion = TargetPhpVersion::create(null);
+        $fileParser = FileParserFactory::createFileParser($targetPhpVersion);
+
+        $classSet = ClassSet::fromDir($input->getOption('from-dir'));
+        foreach ($classSet as $file) {
+            $fileParser->parse($file->getContents(), $file->getRelativePathname());
+            $parsedErrors = $fileParser->getParsingErrors();
+
+            foreach ($parsedErrors as $parsedError) {
+                // TODO qua ce ne vogliamo fare qualcosa?
```
Comment (AlessandroMinoccheri) [inline src/CLI/Command/DebugExpression.php:58]: We can print an error at the end of the process maybe at the moment

File: src/CLI/Command/DebugExpression.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                InputOption::VALUE_REQUIRED,
+                'The folder in which to search the classes',
+                '.'
+            );
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        // TODO come gestiamo le diverse versioni di PHP? Con una opzione?
+        $targetPhpVersion = TargetPhpVersion::create(null);
```
Comment (AlessandroMinoccheri) [inline src/CLI/Command/DebugExpression.php:49]: we can use the same option `target-php-version` like other commands

File: src/CLI/Command/DebugExpression.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            $ruleName = $input->getArgument('expression');
+            $ruleFQCN = 'Arkitect\Expression\ForClasses\\'.$ruleName;
+            $arguments = $input->getArgument('arguments');
+
+            $rule = new $ruleFQCN(...$arguments);
+
+            foreach ($fileParser->getClassDescriptions() as $classDescription) {
+                $violations = new Violations();
+                $rule->evaluate($classDescription, $violations, '');
+                if (0 == $violations->count()) {
```
Comment (AlessandroMinoccheri) [inline src/CLI/Command/DebugExpression.php:70]: Can we use `===`?

Comment (fain182) [reply src/CLI/Command/DebugExpression.php]: Done :slightly_smiling_face:

File: src/CLI/Command/DebugExpression.php (hunk 1 above)
Comment (fain182) [reply src/CLI/Command/DebugExpression.php]: I print it at the beginning, grouped under a warning

File: README.md [hunk 4]
```diff
@@ -96,6 +96,12 @@ return static function (Config $config): void {
 
 # Available rules
 
+**Hint**: If you want to test how a Rule work, you can use the command like `phparkitect debug:expression <RuleName> <arguments>` to check which class satisfy the rule in your current folder.
+
+For example: `phparkitect debug:expression ResideInOneOfTheseNamespaces App`
+
```
Comment (AlessandroMinoccheri) [inline README.md:102]: What do you think about adding a screenshot of the output is too much for this file?

File: src/CLI/Command/DebugExpression.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+            try {
+                $expressionReflection = new \ReflectionClass($ruleFQCN);
+            } catch (\ReflectionException $exception) {
+                $output->writeln("Error: Expression '$ruleName' not found.");
+
+                return 2;
+            }
+
+            $constructorReflection = $expressionReflection->getConstructor();
```
Comment (AlessandroMinoccheri) [inline src/CLI/Command/DebugExpression.php:86]: We can move these checks for arguments in a specific function to keep the main function cleaner, what do you think?

**Review by AlessandroMinoccheri — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (fain182) [reply README.md:102]: I think adding some screenshot to the README could be great, but this is not the most interesting part (the output is just a list of FQCN)...
Maybe we could create an issue to add a screenshot or a gif of phparkitect in action?

File: src/CLI/Command/DebugExpression.php (hunk 5 above)
Comment (fain182) [reply src/CLI/Command/DebugExpression.php]: Good catch, I extracted a method for this part and a couple of others, to simplify this huge method...

File: README.md (hunk 4 above)
Comment (AlessandroMinoccheri) [reply README.md:102]: yes, I think that we could add screenshots in another PR!

**Review by AlessandroMinoccheri — APPROVED** [review]

**Review by micheleorselli — APPROVED** [review]
Great job Pietro 💯

**Pull request merged**